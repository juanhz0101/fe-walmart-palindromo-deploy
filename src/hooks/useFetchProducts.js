import { useEffect, useState } from "react"
import { getProducts, getProductsQuery } from '../helpers/getProducts';

export  const useFetchProducts = ({query = '', pagination = '/products?page=1&limit=20'} ) => {

    const [state, setState] = useState({
        data: [],
        loading: true,
        paginationInfo: {}
    });

    useEffect( () => {

        if (query === '') {
            getProducts(pagination)
            .then( ({ products, paginationInfo }) => {
                setState({
                    data: products,
                    loading: false,
                    paginationInfo
                })
            })
        }else{
            getProductsQuery(query, pagination)
            .then( ({ products, paginationInfo }) => {
                setState({
                    data: products,
                    loading: false,
                    paginationInfo
                })
            })
        }

    }, [ query, pagination])

    return state

}