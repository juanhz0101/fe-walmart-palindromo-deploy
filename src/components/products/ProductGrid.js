import React from 'react'
import { useFetchProducts } from '../../hooks/useFetchProducts'
import ProductGridItem from './ProductGridItem';
import PropTypes from 'prop-types';


function ProductGrid({ query, setQuery }) {
    
    const { data:products, loading, paginationInfo } = useFetchProducts( query );

    const heandleButtonPreviousPage = (e) => ( setQuery(q => ({ ...q, 'pagination': paginationInfo.previous })) )
    const heandleButtonNextPage = (e) => ( setQuery(q => ({ ...q, 'pagination': paginationInfo.next })) )
    const heandleButtonFirstPage = (e) => ( setQuery(q => ({ ...q, 'pagination': paginationInfo.first })) )
    const heandleButtonLastPage = (e) => ( setQuery(q => ({  ...q, 'pagination': paginationInfo.last })) )

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{ query.query }</h3>
            { loading && <p className="animate__animated animate__flash">Loading</p>}

            <div className="card-grid">
                {
                    products.map( ( product ) =>  (
                    
                        <ProductGridItem 
                            key={product.id} 
                            { ...product} 
                            />
                    )) 
                }
            </div>

            <div >
                <p>Total registros: {paginationInfo.totalResources}</p>
                <p>Total registros por pagina: {paginationInfo.limit}</p>
                <p>Total paginas: {paginationInfo.totalPages}</p>
                <p>Pagina actual: {paginationInfo.page}</p>
                <br />
                <p>
                    <button disabled={paginationInfo.previous === null} onClick={ heandleButtonPreviousPage }>Previa</button>
                    
                    <button disabled={paginationInfo.next === null} onClick={ heandleButtonNextPage }>Siguiente</button>

                    <button disabled={paginationInfo.first === null} onClick={ heandleButtonFirstPage }>Primera</button>

                    <button disabled={paginationInfo.last === null} onClick={ heandleButtonLastPage }>Ultima</button>
                </p>
            </div>

        </>
        
    )
}

export default ProductGrid

ProductGrid.propTypes = {
    query: PropTypes.object.isRequired,
    setQuery: PropTypes.func.isRequired
}

