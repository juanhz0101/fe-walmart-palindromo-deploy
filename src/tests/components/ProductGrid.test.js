import React from 'react';
import '@testing-library/jest-dom';
import { shallow} from 'enzyme';
import ProductGrid from "../../components/products/ProductGrid";
import { useFetchProducts } from '../../hooks/useFetchProducts';
jest.mock('../../hooks/useFetchProducts')


describe('Pruebas en <ProductGrid />', () => {

    const query = { query: '', pagination:'/products?page=1&limit=20'};
    const setQuery = jest.fn();

    test('debe de mostrar <ProductGrid /> correctamente', () => {

        
        useFetchProducts.mockReturnValue({
            data: [],
            loading: true,
            paginationInfo:{
                totalResources: 21
            }
        });
        

        const wrapper = shallow( <ProductGrid query={query} setQuery={setQuery} />);
        expect( wrapper ).toMatchSnapshot();
    })

    test('debe de mostrar items cuando se cargan productos useFetchProducts', () => {

        const gifs = [
            {
                id: "2", 
                brand: "dsaasd", 
                description: "zlrwax bñyrh", 
                image: "https://www.lider.cl/catalogo/images/babyIcon.svg", 
                price: "130173", 
                priceWithDiscount: "65086.5", 
                discount: "50"
            },
            {
                id: "532", 
                brand: "dsaasd", 
                description: "zlrwax bñyrh", 
                image: "https://www.lider.cl/catalogo/images/babyIcon.svg", 
                price: "130173", 
                priceWithDiscount: "65086.5", 
                discount: "50"
            }
        ]

        useFetchProducts.mockReturnValue({
            data: gifs,
            loading: false,
            paginationInfo:{
                totalResources: 21
            }
        });
        
        const wrapper = shallow( <ProductGrid query={query} setQuery={setQuery} />);

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('ProductGridItem').length ).toBe( gifs.length )
    })
    

    
})