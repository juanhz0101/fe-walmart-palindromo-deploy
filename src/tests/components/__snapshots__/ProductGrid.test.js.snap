// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ProductGrid /> debe de mostrar <ProductGrid /> correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  />
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
  <div>
    <p>
      Total registros: 
      21
    </p>
    <p>
      Total registros por pagina: 
    </p>
    <p>
      Total paginas: 
    </p>
    <p>
      Pagina actual: 
    </p>
    <br />
    <p>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Previa
      </button>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Siguiente
      </button>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Primera
      </button>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Ultima
      </button>
    </p>
  </div>
</Fragment>
`;

exports[`Pruebas en <ProductGrid /> debe de mostrar items cuando se cargan productos useFetchProducts 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  />
  <div
    className="card-grid"
  >
    <ProductGridItem
      brand="dsaasd"
      description="zlrwax bñyrh"
      discount="50"
      id="2"
      image="https://www.lider.cl/catalogo/images/babyIcon.svg"
      key="2"
      price="130173"
      priceWithDiscount="65086.5"
    />
    <ProductGridItem
      brand="dsaasd"
      description="zlrwax bñyrh"
      discount="50"
      id="532"
      image="https://www.lider.cl/catalogo/images/babyIcon.svg"
      key="532"
      price="130173"
      priceWithDiscount="65086.5"
    />
  </div>
  <div>
    <p>
      Total registros: 
      21
    </p>
    <p>
      Total registros por pagina: 
    </p>
    <p>
      Total paginas: 
    </p>
    <p>
      Pagina actual: 
    </p>
    <br />
    <p>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Previa
      </button>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Siguiente
      </button>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Primera
      </button>
      <button
        disabled={false}
        onClick={[Function]}
      >
        Ultima
      </button>
    </p>
  </div>
</Fragment>
`;
